--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1537,7 +_,7 @@
             return;
         }
 
-        this.getHandle().connection.send(level.clockManager().createFullSyncPacket(this.getHandle()));
+        this.getHandle().connection.send(level.clockManager().createFullSyncPacket());
     }
 
     @Override
