From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AnxeloK <95703440+AnxeloK@users.noreply.github.com>
Date: Tue, 11 Aug 2026 09:08:40 +0300
Subject: [PATCH] Publish plugin enabled state across owners


diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index 0f3d3b757fbb034519be7ecf0f4ed18297a0c443..0645ca447d56f58e97370497027dfe9e4f72994e 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -41,7 +41,7 @@ import org.jspecify.annotations.Nullable;
  */
 @NullMarked
 public abstract class JavaPlugin extends PluginBase {
-    private boolean isEnabled = false;
+    private volatile boolean isEnabled = false;
     private PluginLoader loader = null;
     private Server server = null;
     private File file = null;
