--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1306,7 +_,7 @@
         }
 
         this.console.addLevel(serverLevel);
-        this.console.initWorld(serverLevel, creator);
+        this.console.initWorld(serverLevel);
 
         serverLevel.setSpawnSettings(true);
 
@@ -2928,7 +_,7 @@
 
     @Override
     public int getCurrentTick() {
-        return net.minecraft.server.MinecraftServer.currentTick;
+        return net.minecraft.server.MinecraftServer.getServer().getTickCount();
     }
 
     @Override
