From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Tue, 9 Jun 2026 19:47:51 +0000
Subject: [PATCH] Tune Loom implicit worker default


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
index b49578de93385bd4b5c3eb0a22f949201e1e8685..5b8bd9184e4b5a98dc2e7f874a3fa3fec978c231 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
@@ -74,7 +74,7 @@ public final class MoonriseCommon {
     }
 
     static int loomWorkerThreads(final int totalCores) {
-        return Math.max(1, totalCores / 8);
+        return Math.max(1, Math.min(3, totalCores / 4));
     }
 
     public static final long IO_QUEUE_HOLD_TIME = (long)(25.0e6); // 25ms
diff --git a/src/test/java/ca/spottedleaf/moonrise/common/util/MoonriseCommonTest.java b/src/test/java/ca/spottedleaf/moonrise/common/util/MoonriseCommonTest.java
index f88cd3f3376130259a7e4ef9f3caaa6f7e83d7f3..b8658626b3a7a5e5fa1739dbd294e50e71b3a72d 100644
--- a/src/test/java/ca/spottedleaf/moonrise/common/util/MoonriseCommonTest.java
+++ b/src/test/java/ca/spottedleaf/moonrise/common/util/MoonriseCommonTest.java
@@ -12,8 +12,9 @@ class MoonriseCommonTest {
         Assertions.assertEquals(1, MoonriseCommon.loomWorkerThreads(1));
         Assertions.assertEquals(1, MoonriseCommon.loomWorkerThreads(2));
         Assertions.assertEquals(1, MoonriseCommon.loomWorkerThreads(6));
-        Assertions.assertEquals(1, MoonriseCommon.loomWorkerThreads(12));
-        Assertions.assertEquals(2, MoonriseCommon.loomWorkerThreads(16));
-        Assertions.assertEquals(4, MoonriseCommon.loomWorkerThreads(32));
+        Assertions.assertEquals(2, MoonriseCommon.loomWorkerThreads(8));
+        Assertions.assertEquals(3, MoonriseCommon.loomWorkerThreads(12));
+        Assertions.assertEquals(3, MoonriseCommon.loomWorkerThreads(16));
+        Assertions.assertEquals(3, MoonriseCommon.loomWorkerThreads(32));
     }
 }
