From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Wed, 3 Jun 2026 17:19:50 +0000
Subject: [PATCH] Tighten initial player chunk send burst cap


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index 0ccb2c14548e253d02f3e03e57c529b6b0899ed3..47d6510a68f3a704712d4eba5bd466139f30a9f6 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -429,7 +429,7 @@ public final class RegionizedPlayerChunkLoader {
 
         private static final long MAX_RATE = 10_000L;
         private static final int LOOM_PLAYER_CHUNK_SEND_MAX_PER_TICK = getIntProperty(
-            "Loom.PlayerChunkSendMaxPerTick", 8
+            "Loom.PlayerChunkSendMaxPerTick", 4
         );
         private static final long LOOM_KEEPALIVE_SOFT_BACKPRESSURE_NANOS = getMillisPropertyNanos(
             "Loom.PlayerChunkSendKeepAliveSoftBackpressureMillis", 5_000L
