From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Wed, 10 Jun 2026 15:05:00 +0000
Subject: [PATCH] Restore primed TNT tick counter reset


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index a00d52729d405d7ef3f8ec166a8e7d42b8908bd4..591511d4b2ea65425c8ed64cac087dad6f7b1ad1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -944,6 +944,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 loomDiagnostics.record(LoomTickSectionDiagnostics.BLOCK_ENTITIES, System.nanoTime() - loomBlockEntitiesStart);
             }
             profiler.pop();
+            this.spigotConfig.currentPrimedTnt = 0; // Spigot
         }
 
         loomSectionStart = loomDiagnostics == null ? 0L : System.nanoTime();
