From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Sat, 11 Jul 2026 00:42:34 +0000
Subject: [PATCH] Trim entity tick allocations


diff --git a/net/minecraft/world/entity/InsideBlockEffectApplier.java b/net/minecraft/world/entity/InsideBlockEffectApplier.java
index fa7ae9b80fe0035aca42d18c5983b7f3861bc098..dc642dc329bc0a7996b8b2a71122e1094711e1a5 100644
--- a/net/minecraft/world/entity/InsideBlockEffectApplier.java
+++ b/net/minecraft/world/entity/InsideBlockEffectApplier.java
@@ -53,30 +53,36 @@ public interface InsideBlockEffectApplier {
         public void applyAndClear(final Entity entity) {
             this.flushStep();
 
-            for (Consumer<Entity> effect : this.finalEffects) {
-                if (!entity.isAlive()) {
-                    break;
-                }
+            if (!this.finalEffects.isEmpty()) {
+                for (Consumer<Entity> effect : this.finalEffects) {
+                    if (!entity.isAlive()) {
+                        break;
+                    }
 
-                effect.accept(entity);
+                    effect.accept(entity);
+                }
+                this.finalEffects.clear();
             }
 
-            this.finalEffects.clear();
             this.lastStep = -1;
         }
 
         private void flushStep() {
             for (InsideBlockEffectType type : APPLY_ORDER) {
                 List<Consumer<Entity>> beforeEffects = this.beforeEffectsInStep.get(type);
-                this.finalEffects.addAll(beforeEffects);
-                beforeEffects.clear();
+                if (!beforeEffects.isEmpty()) {
+                    this.finalEffects.addAll(beforeEffects);
+                    beforeEffects.clear();
+                }
                 if (this.effectsInStep.remove(type) instanceof final Consumer<Entity> recordedEffect) { // Paper - track position inside effect was triggered on - better than null check to avoid diff.
                     this.finalEffects.add(recordedEffect); // Paper - track position inside effect was triggered on
                 }
 
                 List<Consumer<Entity>> afterEffects = this.afterEffectsInStep.get(type);
-                this.finalEffects.addAll(afterEffects);
-                afterEffects.clear();
+                if (!afterEffects.isEmpty()) {
+                    this.finalEffects.addAll(afterEffects);
+                    afterEffects.clear();
+                }
             }
         }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7cb742eb47b10aa8c409e268a47565fb2750670d..1590e57039a5c15f1e041131679d203a5f22a566 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1324,6 +1324,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     private void refreshDirtyAttributes() {
         Set<AttributeInstance> attributesToUpdate = this.getAttributes().getAttributesToUpdate();
+        if (attributesToUpdate.isEmpty()) {
+            return;
+        }
 
         for (AttributeInstance changedAttributeInstance : attributesToUpdate) {
             this.onAttributeUpdated(changedAttributeInstance.getAttribute());
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 75d8953845b2e9acb62b19a859d0e39fd63533eb..6c175046c91079279b69b66f5165c8474460cd57 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -91,7 +91,12 @@ public class GoalSelector {
             }
         }
 
-        this.lockedFlags.entrySet().removeIf(entry -> !entry.getValue().isRunning());
+        for (final Goal.Flag flag : GOAL_FLAG_VALUES) {
+            final WrappedGoal lockedGoal = this.lockedFlags.get(flag);
+            if (lockedGoal != null && !lockedGoal.isRunning()) {
+                this.lockedFlags.remove(flag);
+            }
+        }
         profiler.pop();
         profiler.push("goalUpdate");
 
