From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Sat, 11 Jul 2026 00:42:34 +0000
Subject: [PATCH] Trim entity tick allocations


diff --git a/net/minecraft/world/entity/InsideBlockEffectApplier.java b/net/minecraft/world/entity/InsideBlockEffectApplier.java
index fc7cfc75e3e5c274f01498e4afd8693236b4e394..6347a656f631838841eaaef685c5a931646090c5 100644
--- a/net/minecraft/world/entity/InsideBlockEffectApplier.java
+++ b/net/minecraft/world/entity/InsideBlockEffectApplier.java
@@ -53,30 +53,36 @@ public interface InsideBlockEffectApplier {
         public void applyAndClear(final Entity entity) {
             this.flushStep();
 
-            for (Consumer<Entity> effect : this.finalEffects) {
-                if (!entity.isAlive()) {
-                    break;
-                }
+            if (!this.finalEffects.isEmpty()) {
+                for (Consumer<Entity> effect : this.finalEffects) {
+                    if (!entity.isAlive()) {
+                        break;
+                    }
 
-                effect.accept(entity);
+                    effect.accept(entity);
+                }
+                this.finalEffects.clear();
             }
 
-            this.finalEffects.clear();
             this.lastStep = -1;
         }
 
         private void flushStep() {
             for (InsideBlockEffectType type : APPLY_ORDER) {
                 List<Consumer<Entity>> beforeEffects = this.beforeEffectsInStep.get(type);
-                this.finalEffects.addAll(beforeEffects);
-                beforeEffects.clear();
+                if (!beforeEffects.isEmpty()) {
+                    this.finalEffects.addAll(beforeEffects);
+                    beforeEffects.clear();
+                }
                 if (this.effectsInStep.remove(type) instanceof final Consumer<Entity> recordedEffect) { // Paper - track position inside effect was triggered on - better than null check to avoid diff.
                     this.finalEffects.add(recordedEffect); // Paper - track position inside effect was triggered on
                 }
 
                 List<Consumer<Entity>> afterEffects = this.afterEffectsInStep.get(type);
-                this.finalEffects.addAll(afterEffects);
-                afterEffects.clear();
+                if (!afterEffects.isEmpty()) {
+                    this.finalEffects.addAll(afterEffects);
+                    afterEffects.clear();
+                }
             }
         }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index e23720de0df7759b1be384a4d118962ae6e23483..c9aa469f2b8c9d232c13adc6effeefe1ac1b58e1 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1328,6 +1328,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     private void refreshDirtyAttributes() {
         Set<AttributeInstance> attributesToUpdate = this.getAttributes().getAttributesToUpdate();
+        if (attributesToUpdate.isEmpty()) {
+            return;
+        }
 
         for (AttributeInstance changedAttributeInstance : attributesToUpdate) {
             this.onAttributeUpdated(changedAttributeInstance.getAttribute());
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 0439543fbb25ddb974550fe9dbdaa479e330937d..6f948996b82827222d8612b0c5df0caafa92e523 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -91,7 +91,12 @@ public class GoalSelector {
             }
         }
 
-        this.lockedFlags.entrySet().removeIf(entry -> !entry.getValue().isRunning());
+        for (final Goal.Flag flag : GOAL_FLAG_VALUES) {
+            final WrappedGoal lockedGoal = this.lockedFlags.get(flag);
+            if (lockedGoal != null && !lockedGoal.isRunning()) {
+                this.lockedFlags.remove(flag);
+            }
+        }
         profiler.pop();
         profiler.push("goalUpdate");
 
