From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anxelok <anxelok@localhost>
Date: Sun, 12 Jul 2026 20:36:52 +0000
Subject: [PATCH] Avoid per-packet bundle consumers


diff --git a/net/minecraft/network/PacketBundleUnpacker.java b/net/minecraft/network/PacketBundleUnpacker.java
index e20531becd31cb37b0d1780630d7428ab6dd1380..9f2e09e7a21d591c037ab7d6d3cee24d5964b9bd 100644
--- a/net/minecraft/network/PacketBundleUnpacker.java
+++ b/net/minecraft/network/PacketBundleUnpacker.java
@@ -15,7 +15,7 @@ public class PacketBundleUnpacker extends MessageToMessageEncoder<Packet<?>> {
 
     @Override
     protected void encode(final ChannelHandlerContext ctx, final Packet<?> msg, final List<Object> out) throws Exception {
-        this.bundlerInfo.unbundlePacket(msg, out::add);
+        this.bundlerInfo.unbundlePacket(msg, out);
         if (msg.isTerminal()) {
             ctx.pipeline().remove(ctx.name());
         }
diff --git a/net/minecraft/network/protocol/BundlerInfo.java b/net/minecraft/network/protocol/BundlerInfo.java
index d2488b2648860b8d511103b227c47d7e77651c8e..b7c8af728cd4e14094553d419fe84d9ea5e33d72 100644
--- a/net/minecraft/network/protocol/BundlerInfo.java
+++ b/net/minecraft/network/protocol/BundlerInfo.java
@@ -28,6 +28,20 @@ public interface BundlerInfo {
                 }
             }
 
+            @Override
+            public void unbundlePacket(final Packet<?> packet, final List<Object> output) {
+                if (packet.type() == bundlePacketType) {
+                    P bundlerPacket = (P)packet;
+                    output.add(delimiterPacket);
+                    for (Packet<? super T> subPacket : bundlerPacket.subPackets()) {
+                        output.add(subPacket);
+                    }
+                    output.add(delimiterPacket);
+                } else {
+                    output.add(packet);
+                }
+            }
+
             @Override
             public BundlerInfo.@Nullable Bundler startPacketBundling(final Packet<?> packet) {
                 return packet == delimiterPacket ? new BundlerInfo.Bundler() {
@@ -54,6 +68,10 @@ public interface BundlerInfo {
 
     void unbundlePacket(Packet<?> packet, Consumer<Packet<?>> output);
 
+    default void unbundlePacket(final Packet<?> packet, final List<Object> output) {
+        this.unbundlePacket(packet, output::add);
+    }
+
     BundlerInfo.@Nullable Bundler startPacketBundling(Packet<?> packet);
 
     interface Bundler {
